codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/auth/AuthScheme.java

```diff
@@ -0,0 +1,278 @@
+
+public final class AuthScheme {
+    private final ShapeId shapeId;
+    private final String rubyAuthScheme;
+    private final String rubyIdentityType;
+    private final Map<String, String> additionalAuthParams;
+    private final ClientConfig identityResolverConfig;
+    private final WriteAdditionalFiles writeAdditionalFiles;
+    private final ExtractProperties extractSignerProperties;
+    private final Map<String, String> signerProperties;
+    private final ExtractProperties extractIdentityProperties;
+    private final Map<String, String> identityProperties;
+
+    private AuthScheme(Builder builder) {
+        this.shapeId = builder.shapeId;
+        this.rubyAuthScheme = builder.rubyAuthScheme;
+        this.rubyIdentityType = builder.rubyIdentityType;
+        this.additionalAuthParams = builder.additionalAuthParams;
+        this.identityResolverConfig = builder.identityResolverConfig;
+        this.writeAdditionalFiles = builder.writeAdditionalFiles;
+        this.extractSignerProperties = builder.extractSignerProperties;
+        this.signerProperties = builder.signerProperties;
+        this.extractIdentityProperties = builder.extractIdentityProperties;
+        this.identityProperties = builder.identityProperties;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * @return the shapeId of the auth scheme.
+     */
+    public ShapeId getShapeId() {
+        return shapeId;
+    }
+
+    /**
+     * @return the initialized ruby class of the auth scheme.
+     */
+    public String getRubyAuthScheme() {
+        return rubyAuthScheme;
+    }
+
+    /**
+     * @return the ruby identity type (class).
+     */
+    public String getRubyIdentityType() {
+        return rubyIdentityType;
+    }
+
+    /**
+     * @return additional auth params for the auth scheme.
+     */
+    public Map<String, String> getAdditionalAuthParams() {
+        return additionalAuthParams;
+    }
+
+    /**
+     * @return the ClientConfig for the identity resolver.
+     */
+    public ClientConfig getIdentityResolverConfig() {
+        return identityResolverConfig;
+    }
+
+    /**
+     * Write additional files required by this auth scheme.
+     *
+     * @param context generation context
+     * @return List of additional files written out
+     */
+    public List<String> writeAdditionalFiles(GenerationContext context) {
+        return writeAdditionalFiles.writeAdditionalFiles(context);
+    }
+
+    /**
+     * @param trait the trait to extract Signer properties from.
+     */
+    public void extractSignerProperties(Trait trait) {
+         Map<String, String> signerProperties = extractSignerProperties.extractProperties(trait);
+         signerProperties.forEach((key, value) -> this.signerProperties.put(key, value));
+    }
+
+    /**
+     * @return the Signer properties for the auth scheme.
+     */
+    public Map<String, String> getSignerProperties() {
+        return signerProperties;
```

Comment:
It seems like this method should take a Trait and pass it to the extractSignerProperties to avoid mutating authScheme.